### ISSUE TITLE
Unit parsing errors produce misleading advice in certain cases

### ISSUE THREAD

**Issue by weaverba137** [astropy/astropy#15313]
### Description

When certain units cannot be parsed, a specific message is printed that can be misleading in certain cases. If a user follows the advice, the result is an exception.

### Expected behavior

The warning about defining a new unit in order to allow it to be parsed should take into account units that are already defined. In addition, `astropy.units` should go ahead and parse the unit anyway, if it is already defined in `astropy.units`.

### How to Reproduce

1. Set up a FITS file.
```python
>>> import numpy as np
>>> import astropy.units as u
>>> from astropy.io import fits
>>> from astropy.table import Table, QTable
>>> rng = np.random.default_rng()
>>> targetid = rng.integers(0, 2**64, (20, ), dtype=np.uint64)
>>> mag = 1e9 * rng.random((20, ), dtype=np.float32)
>>> columns = fits.ColDefs([fits.Column(name='TARGETID', format='K', bzero=2**63, array=targetid),
>>>                         fits.Column(name='FLUX', format='E', array=mag)])
>>> hdu0 = fits.PrimaryHDU()
>>> hdu0.header.append(('EXTNAME', 'PRIMARY', 'extension name'))
>>> hdu0.add_checksum()
>>> hdu1 = fits.BinTableHDU.from_columns(columns, name='BINTABLE', uint=True)
>>> hdu1.header.insert('TFORM2', ('TUNIT2', 'nanomaggy', 'Units for FLUX'), after=True)
>>> hdu1.add_checksum()
>>> hdulist = fits.HDUList([hdu0, hdu1])
>>> hdulist.writeto('demonstrate_maggy.fits', overwrite=True)
```
2. Try to read the FITS file.
```python
>>> t = Table.read('demonstrate_maggy.fits', hdu='BINTABLE')
WARNING: UnitsWarning: 'nanomaggy' did not parse as fits unit: At col 0, Unit 'nanomaggy' not supported by the FITS standard.  If this is meant to be a custom unit, define it with 'u.def_unit'. To have it recognized inside a file reader or other code, enable it with 'u.add_enabled_units'. For details, see https://docs.astropy.org/en/latest/units/combining_and_defining.html [astropy.units.core]
```
3. Great, I'll define the unit then!
```python
>>> nanomaggy = u.def_unit('nanomaggy', 1.0e-9 * 3631*u.Jy)
>>> u.add_enabled_units( [nanomaggy,])
Traceback (most recent call last):
  File "/Users/weaver/Documents/Code/git/desihub/desiutil/demonstrate_maggy.py", line 27, in <module>
    u.add_enabled_units( [nanomaggy,])
  File "/Users/weaver/Documents/local/products/venv/desiconda/lib/python3.10/site-packages/astropy/units/core.py", line 469, in add_enabled_units
    get_current_unit_registry().add_enabled_units(units)
  File "/Users/weaver/Documents/local/products/venv/desiconda/lib/python3.10/site-packages/astropy/units/core.py", line 221, in add_enabled_units
    raise ValueError(
ValueError: Object with name 'nanomaggy' already exists in namespace. Filter the set of units to avoid name clashes before enabling them.
```
4. OK, so if `nanomaggy` is already defined, why didn't you say so?  I can understand the FITS warning, but `astropy.units` clearly has the capability to parse it anyway.

### Versions

```python
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import astropy; print("astropy", astropy.__version__)
import numpy; print("Numpy", numpy.__version__)
import erfa; print("pyerfa", erfa.__version__)
import scipy; print("Scipy", scipy.__version__)
import matplotlib; print("Matplotlib", matplotlib.__version__)
```

```
macOS-13.5.2-x86_64-i386-64bit
Python 3.10.10 (main, Feb 10 2023, 08:40:50) [Clang 14.0.0 (clang-1400.0.29.202)]
astropy 5.2.1
Numpy 1.22.4
pyerfa 2.0.0.1
Scipy 1.8.1
Matplotlib 3.6.2
```

**Comment by weaverba137** [comment]
PS, let me emphasize that in the example above, the table *is* read successfully, and the resulting column *does* have units of `nanomaggy`.  The only concern here is the warning message that is printed, and its consequences if a user follows the advice.

**Comment by pllim** [comment]
If you turn the warning into exception, the warning is from here:

https://github.com/astropy/astropy/blob/c095956a927bd751cb10df36ba2b7ab00068ffa3/astropy/io/fits/connect.py#L290

You can reproduce it like this:

```python
u.Unit('nanomaggy', format="fits")
```

There is no warning if you don't force `format="fits"`. I am not familiar with the FITS standards. Is "nanomaggy" not acceptable by FITS?

Trying to silence the warning is not going to help either:

```python
>>> u.Unit('nanomaggy', format="fits", parse_strict="silent")
UnrecognizedUnit(nanomaggy)
>>> u.Unit('nanomaggy')
Unit("nmgy")
```

**Comment by weaverba137** [comment]
Correct, `nanomaggy` is not part of the FITS standard. Nevertheless, it is defined by `astropy.units.photometric`.  I *do* think it is fair to issue a warning about a unit not being in the FITS standard.

However, it's the second part of the warning that is misleading here, and that's why I'm reporting this.

**Comment by pllim** [comment]
Thanks!

**Comment by pllim** [comment]
Side note: When I pass `hdu1` directly to `Table`, I don't see that warning. 🤷

**Comment by weaverba137** [comment]
Interesting. I also want to test what happens with some other unit that isn't already defined, I'll give that a try tomorrow.

**Comment by weaverba137** [comment]
An additional test:

1. I created a FITS file exactly like the example above except I replaced `nanomaggy` with `nanofluxy`.
2. *Before* attempting to read the FITS file at all, I added `nanofluxy` to `astropy.units`:
```python
>>> nanofluxy = u.def_unit('nanofluxy', 1.0e-9*u.Jy)
>>> u.add_enabled_units( [nanofluxy,])
>>> t = Table.read('demonstrate_maggy.fits', hdu='BINTABLE')
```
3. I get the same warning message. That's not surprising, because [the warning message is hard-coded](https://github.com/astropy/astropy/blob/c095956a927bd751cb10df36ba2b7ab00068ffa3/astropy/units/core.py#L2111).
4. For someone who doesn't know that the warning message is hard-coded, then the message *would* be a surprise, *e.g.* "I just defined that unit, why am I still getting this message about the unit being undefined?"

**Comment by pllim** [comment]
I guess the "connect" logic in `io.fits` to `Table` could catch this warning and and throw another warning that is more useful? 🤔

**Comment by weaverba137** [comment]
Still, if a quick workaround is needed, maybe that's the way it has to be.

**Comment by pllim** [comment]
> https://github.com/astropy/astropy/issues/15313#issuecomment-1721862790

Hmm, maybe, but...

* such warning is very specific to FITS format, so do we want that check so deep in the core?
* and why does the warning not get thrown when HDU is passed instead of filename? Such inconsistent may suggest a bug in the FITS connect function.

**Comment by weaverba137** [comment]
So you suggest removing the warning from `astropy.units` entirely?  It certainly seems reasonable that warnings pertaining to FITS-specific units should actually be in `io.fits`.

I did not realize there was a difference between passing a filename and passing a HDU, that's definitely weird!

**Comment by pllim** [comment]
No, I mean don't touch the `units` core functionality but rather do FITS specific warning handling within `io.fits` connect code. But then again, I am not the maintainer of either subpackages, so maybe @saimn or @mhvk have better ideas. 🤞

**Comment by weaverba137** [comment]
OK, got it.

**Comment by mhvk** [comment]
I only remember that the previous time we dealt with this was mostly to ensure we gave people a hint on how to define custom units to solve a unit not being recognized. Since units is a base for FITS, ideally nothing in it is all that FITS aware, so I think the problem here is better solved in `io.fits`. But I don't have the overview of how things interact to tell what would be the best place to tackle this...

**Comment by neutrinoceros** [comment]
Taking a look now.
Let me delve into the secondary bug at first
(I'm running a very slightly simplified and self-contained version of the MWE, reproduced hereafter for convenience)
```python
import numpy as np
import astropy.units as u
from astropy.table import Table
from astropy.io import fits

rng = np.random.default_rng()
targetid = rng.integers(0, 2**64, (1, ), dtype=np.uint64)
mag = 1e9 * rng.random((1,), dtype=np.float32)
columns = fits.ColDefs([fits.Column(name='TARGETID', format='K', bzero=2**63, array=targetid),
                        fits.Column(name='FLUX', format='E', array=mag)])
hdu0 = fits.PrimaryHDU()
hdu0.header.append(('EXTNAME', 'PRIMARY', 'extension name'))
hdu0.add_checksum()
hdu1 = fits.BinTableHDU.from_columns(columns, name='BINTABLE', uint=True)
hdu1.header.insert('TFORM2', ('TUNIT2', 'nanomaggy', 'Units for FLUX'), after=True)
hdu1.add_checksum()
hdulist = fits.HDUList([hdu0, hdu1])
hdulist.writeto('demonstrate_maggy.fits', overwrite=True)

t1 = Table.read(hdu1, hdu='BINTABLE')
t2 = Table.read('demonstrate_maggy.fits', hdu='BINTABLE')
```

printing `t1` and `t2` reveals that `t1` seems to have lost units metadata:
```
>>> print(t1)
      TARGETID           FLUX   
       uint64          float32  
-------------------- -----------
11049962027143826759 743222300.0

>>> print(t2)
      TARGETID           FLUX   
                      nanomaggy 
       uint64          float32  
-------------------- -----------
11049962027143826759 743222300.0
```

so I guess that's another symptom of the suspected bug, and maybe more easily exploitable than the difference between getting a warning or not.
At a high level, the difference is that, when passed a filename, `read_table_fits` will first read it as a `HDUList`, then recurse
https://github.com/astropy/astropy/blob/48a792f9d8dc659f428e9c6f0e9e9964a1b60682/astropy/io/fits/connect.py#L234-L252
In the nested call, the following special-case code will be run https://github.com/astropy/astropy/blob/48a792f9d8dc659f428e9c6f0e9e9964a1b60682/astropy/io/fits/connect.py#L184
None of this happens if you just pass in a `BinTableHDU`, as is evident from the first listing.
Deeper investigation is needed, but I think this bug (and discussion) deserves its own issue, so I'll open a separate one.
I also intend to come back and study the main bug reported here more closely.
